src/Analyser/MutatingScope.php

```diff
@@ -4514,17 +4513,14 @@ private static function generalizeType(Type $a, Type $b): Type
 		);
 	}
 
-	private static function getArrayDepth(ArrayType $type): int
+	private static function getArrayDepth(Type $type): int
 	{
 		$depth = 0;
-		while ($type instanceof ArrayType) {
+		$arrays = TypeUtils::getAnyArrays($type);
+		while (count($arrays) > 0) {
 			$temp = $type->getIterableValueType();
-			$arrays = $temp->getArrays();
-			if (count($arrays) === 1) {
-				$type = $arrays[0];
-			} else {
-				$type = $temp;
-			}
+			$type = $temp;
+			$arrays = TypeUtils::getAnyArrays($type);
```

Comment:
I think our problem is that `getAnyArrays` does not return arrays for the [types in question](https://github.com/phpstan/phpstan-src/pull/2134/files#r1059638126) because we are working with a union of intersection types

![grafik](https://user-images.githubusercontent.com/120441/210138937-abdd8966-5834-4378-bfab-926dfbb6ba40.png)

and `TypeUtils::getAnyArrays` does not recursive checks of intersections within unions